src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketRequestExecutor.java

```diff
@@ -0,0 +1,145 @@
+        this.objectMapper = objectMapper;
+        this.credentials = credentials;
+    }
+
+    /**
+     * Returns the root URL of Bitbucket server.
+     */
+    public HttpUrl getBaseUrl() {
+        return bitbucketBaseUrl;
+    }
+
+    /**
+     * Provide the base Rest path
+     *
+     * @return HttpUrl of the core rest path;
+     */
+    public HttpUrl getCoreRestPath() {
+        return coreRestPathUrl;
+    }
+
+    /**
+     * Make a GET request to the url given. This method will add authentication headers as needed.
+     * If the requested resource is paged, or the return type is generified use this method,
+     * otherwise the {@link #makeGetRequest(HttpUrl, Class)} is most likely a better choice.
+     *
+     * @param url        url to connect to
+     * @param returnType type reference used when getting generified objects (such as pages)
+     * @param <T>        type to return
+     * @return a deserialized object of type T
+     * @see #makeGetRequest(HttpUrl, Class)
+     */
+    public <T> BitbucketResponse<T> makeGetRequest(HttpUrl url, TypeReference<T> returnType) {
+        return makeGetRequest(url, in -> objectMapper.readValue(in, returnType));
+    }
+
+    /**
+     * Make a GET request to the url given. This method will add authentication headers as needed.
+     * <em>Note!</em> this method <em>cannot</em> be used to retrieve entities that makes use of
+     * generics (such as {@link BitbucketPage}) for that use {@link #makeGetRequest(HttpUrl,
+     * TypeReference)} instead.
+     *
+     * @param url        url to connect to
+     * @param returnType class of the desired return type. Do note that if the type is generified
+     *                   this method will not work
+     * @param <T>        type to return
+     * @return a deserialized object of type T
+     * @see #makeGetRequest(HttpUrl, TypeReference)
+     */
+    public <T> BitbucketResponse<T> makeGetRequest(HttpUrl url, Class<T> returnType) {
+        return makeGetRequest(url, in -> objectMapper.readValue(in, returnType));
+    }
+
+    /**
+     * Makes a POST request to the given URL with given request payload.
+     *
+     * @param url,            the URL to make the request to
+     * @param requestPayload, JSON payload which will be marshalled to send it with POST.
+     * @param returnType,     Class of expected return type
+     * @param <T>,            Type of Request payload.
+     * @param <R>,            Return type
+     * @return
+     */
+    public <T, R> BitbucketResponse<R> makePostRequest(HttpUrl url, T requestPayload, Class<R> returnType) {
+        ObjectReader<R> reader = in -> objectMapper.readValue(in, returnType);
+        return httpRequestExecutor.executePost(url, credentials, marshall(requestPayload), response ->
+                new BitbucketResponse<>(response.headers().toMultimap(), unmarshall(reader, response.body())));
+    }
+
+    private void ensureNonEmptyBode(Response response) {
```

Comment:
```suggestion
    private void ensureNonEmptyBody(Response response) {
```
typo? I propose we name it ensureNonEmptyBod**y**

Reply:
Actually, I think we should just remove this method. You call it in one spot, and in another you just blindly unmarshall the body which would result in a NPE when the body is null. 
How about we change the `marshall` and `unmarshall` to accept `null` and throw a ` throw new NoContentException(`
if that happens.

Reply:
I have fixed the type and handled null in marshall and unmarshall. We were logging response code as well and I did not want to send this to these methods.